File: 백준/Gold/8972. 미친 아두이노/미친 아두이노.py
import sys
input = sys.stdin.readline

#보드를 벗어나는 입력은 주어지지 않는다.
def move_arduino(r, c):
    min_dist = 201
    result = (r, c)

    for i in range(1, 10):
        dr = r + direction_y[i]
        dc = c + direction_x[i]

        dist = calc_distance(I_pos[0], I_pos[1], dr, dc)

        if (min_dist > dist):
            result = (dr, dc)
            min_dist = dist
            
    return result

def calc_distance(r1, c1, r2, c2):
    return abs(r1-r2) + abs(c1-c2)

direction_x = [0, -1, 0, 1, -1, 0, 1, -1, 0, 1]
direction_y = [0, 1, 1, 1, 0, 0, 0, -1, -1, -1]

R, C = map(int, input().split())
I_pos = [0, 0]
arduino_pos_set = set()

for r in range(R):
    input_data = input().rstrip()
    for c, data in enumerate(input_data):
        if (data == 'I'):
            I_pos = [r, c]
        if (data == 'R'):
            arduino_pos_set.add((r, c))

move_directions = list(map(int, list(map(str, input().rstrip()))))

for cnt, dir in enumerate(move_directions):
    new_arduino_pos_set = set()
    duplicated_pos_set = set()

    # move I
    I_pos = [I_pos[0] + direction_y[dir], I_pos[1] + direction_x[dir]]

    # move arduinos
    for r, c in arduino_pos_set:
        moved_pos = move_arduino(r, c)

        if (I_pos[0] == moved_pos[0] and I_pos[1] == moved_pos[1]):
            print("kraj", cnt+1)
            exit(0)

        if (moved_pos in new_arduino_pos_set):
            duplicated_pos_set.add(moved_pos)
        else:
            new_arduino_pos_set.add(moved_pos)
        
    arduino_pos_set = new_arduino_pos_set.difference(duplicated_pos_set)

answer = [['.' for _ in range(C)] for _ in range(R)]
answer[I_pos[0]][I_pos[1]] = 'I'
for r, c in arduino_pos_set:
    answer[r][c] = 'R'

for ans in answer:
    print("".join(ans))

File: 백준/Gold/8972. 미친 아두이노/미친 아두이노.js
const fs = require("fs");
const filePath = process.platform === "linux" ? "/dev/stdin" : "input.txt";
const input = fs.readFileSync(filePath).toString().trim().split("\n");

function differenceSet(set1, set2) {
    return new Set([...set1].filter((item) => !set2.has(item)));
}

function moveArduino(r, c) {
    let minDistance = 201;
    let result = [r, c];

    for (let i = 1; i < 10; i++) {
        const dr = r + directionY[i];
        const dc = c + directionX[i];

        const distance = calcDistance(IPos[0], IPos[1], dr, dc);

        if (minDistance > distance) {
            minDistance = distance;
            result = [dr, dc];
        }
    }
    return getIndex(result[0], result[1]);
}

function calcDistance(r1, c1, r2, c2) {
    return Math.abs(r1 - r2) + Math.abs(c1 - c2);
}

function getIndex(r, c) {
    return r * C + c;
}

function getPos(index) {
    return [Math.floor(index / C), index % C];
}

let [R, C] = input[0].split(" ").map(Number);
let idx = 1;
let IPos = [];
const directionX = [0, -1, 0, 1, -1, 0, 1, -1, 0, 1];
const directionY = [0, 1, 1, 1, 0, 0, 0, -1, -1, -1];
let arduinoPosSet = new Set();

for (let r = 0; r < R; r++) {
    inputData = input[idx++].trim();

    for (let c = 0; c < C; c++) {
        if (inputData[c] === "I") {
            IPos = [r, c];
        } else if (inputData[c] === "R") {
            arduinoPosSet.add(getIndex(r, c));
        }
    }
}

const moveDirections = input[idx++].split("").map(Number);
for (let cnt = 0; cnt < moveDirections.length; cnt++) {
    const newArduinoPosSet = new Set();
    const duplicatedPosSet = new Set();

    IPos[0] = IPos[0] + directionY[moveDirections[cnt]];
    IPos[1] = IPos[1] + directionX[moveDirections[cnt]];
    const IPosIndex = getIndex(IPos[0], IPos[1]);

    for (const idx of arduinoPosSet) {
        const arduinoPos = getPos(idx);
        const movedPosIndex = moveArduino(arduinoPos[0], arduinoPos[1]);

        if (IPosIndex === movedPosIndex) {
            console.log("kraj", cnt + 1);
            process.exit();
        }

        if (newArduinoPosSet.has(movedPosIndex)) {
            duplicatedPosSet.add(movedPosIndex);
        } else {
            newArduinoPosSet.add(movedPosIndex);
        }
    }

    arduinoPosSet = differenceSet(newArduinoPosSet, duplicatedPosSet);
}

let answer = Array.from({ length: R }, () => Array.from({ length: C }, () => "."));
answer[IPos[0]][IPos[1]] = "I";

for (const idx of arduinoPosSet) {
    const arduinoPos = getPos(idx);
    answer[arduinoPos[0]][arduinoPos[1]] = "R";
}

for (const ans of answer) {
    console.log(ans.join(""));
}
